api/src/main/java/org/infinispan/api/async/AsyncCache.java

```diff
@@ -0,0 +1,186 @@
+package org.infinispan.api.async;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow;
+
+import org.infinispan.api.common.CacheEntry;
+import org.infinispan.api.common.events.cache.CacheContinuousQueryEvent;
+import org.infinispan.api.common.events.cache.CacheEntryEvent;
+import org.infinispan.api.common.events.cache.CacheEntryEventType;
+import org.infinispan.api.common.events.cache.CacheListenerOptions;
+import org.infinispan.api.common.tasks.EntryConsumerTask;
+import org.infinispan.api.configuration.CacheConfiguration;
+
+/**
+ * @since 14.0
+ **/
+public interface AsyncCache<K, V> {
+   /**
+    * The name of the cache.
+    *
+    * @return
+    */
+   String name();
+
+   /**
+    * The configuration for this cache.
+    *
+    * @return
+    */
+   CompletionStage<CacheConfiguration> configuration();
+
+   /**
+    * Return the container of this cache
+    * @return
+    */
+   AsyncContainer container();
+
+   /**
+    * Get the value of the Key if such exists
+    *
+    * @param key
+    * @return the value
+    */
+   CompletionStage<V> get(K key);
+
+   /**
+    * Insert the key/value if such key does not exist
+    *
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<Boolean> putIfAbsent(K key, V value);
+
+   /**
+    * @param key
+    * @param value
+    * @return Void
+    */
+   CompletionStage<V> put(K key, V value);
+
+   /**
+    * @param key
+    * @param value
+    * @return
+    */
+   CompletionStage<Void> set(K key, V value);
+
+   /**
+    * Delete the key
+    *
+    * @param key
+    * @return whether the entry was removed.
+    */
+   CompletionStage<Boolean> remove(K key);
+
+   /**
+    * Removes the key and returns its value if present.
+    *
+    * @param key
+    * @return the value of the key before removal. Returns null if the key didn't exist.
+    */
+   CompletionStage<V> getAndRemove(K key);
+
+   /**
+    * Retrieve all keys
+    *
+    * @return
+    */
+   Flow.Publisher<K> keys();
+
+   /**
+    * Retrieve all entries
+    *
+    * @return
+    */
+   Flow.Publisher<? extends CacheEntry<K, V>> entries();
+
+   /**
+    * @param entries
+    * @return Void
+    */
+   CompletionStage<Void> putAll(Map<K, V> entries);
```

Comment:
Did we want not want streaming methods for these arguments? e.g. `putAll(Flow.Publisher<Map<K, V>>)`